doc/neps/nep-0031-uarray.rst

```diff
@@ -0,0 +1,387 @@
+``__array_function__`` protocol in NEP-18 `[2]`_, but this has shown the need
+for further protocols to be developed, including a protocol for coercion (see
+`[5]`_). The reasons these overrides are needed have been extensively discussed
+in the references, and this NEP will not attempt to go into the details of why
+these are needed. Another pain point requiring yet another protocol is the
+duck-array protocol (see `[9]`_).
+
+This NEP takes a more holistic approach: It assumes that there are parts of
+the API that need to be overridable, and that these will grow over time. It
+provides a general framework and a mechanism to avoid a design of a new
+protocol each time this is required.
+
+This NEP proposes the following: That ``unumpy`` `[8]`_  becomes the
+recommended override mechanism for the parts of the NumPy API not yet covered
+by ``__array_function__`` or ``__array_ufunc__``, and that ``uarray`` is
+vendored into a new namespace within NumPy to give users and downstream
+dependencies access to these overrides.  This vendoring mechanism is similar
+to what SciPy decided to do for making ``scipy.fft`` overridable (see `[10]`_).
+
+
+Detailed description
+--------------------
+
+**Note:** *This section will not attempt to go into too much detail about
+``uarray``, that is the purpose of the ``uarray`` documentation.* `[1]`_
+*However, the NumPy community will have input into the design of
+``uarray``, via the issue tracker.*
+
+``uarray`` Primer
+^^^^^^^^^^^^^^^^^
+
+Defining backends
+~~~~~~~~~~~~~~~~~
+
+``uarray`` consists of two main protocols: ``__ua_convert__`` and
+``__ua_function__``, called in that order, along with ``__ua_domain__``, which
+is a string defining the domain of the backend. If any of the protocols return
+``NotImplemented``, we fall back to the next backend.
+
+``__ua_convert__`` is for conversion and coercion. It has the signature
+``(dispatchables, coerce)``, where ``dispatchables`` is an iterable of
+``ua.Dispatchable`` objects and ``coerce`` is a boolean indicating whether or
+not to force the conversion. ``ua.Dispatchable`` is a simple class consisting
+of three simple values: ``type``, ``value``, and ``coercible``.
+``__ua_convert__`` returns an iterable of the converted values, or
+``NotImplemented`` in the case of failure. Returning ``NotImplemented``
+here will cause ``uarray`` to move to the next available backend.
+
+``__ua_function__`` has the signature ``(func, args, kwargs)`` and defines
+the actual implementation of the function. It recieves the function and its
+arguments. Returning ``NotImplemented`` will cause a move to the default
+implementation of the function if one exists, and failing that, the next
+backend.
+
+If all backends are exhausted, a ``ua.BackendNotImplementedError`` is raised.
+
+Backends can be registered for permanent use if required.
+
+Defining overridable methods
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+To define an overridable method, one needs a few things:
```

Comment:
I think that in Python a *method* is defined as a callable attached to a class, whereas you're talking about top-level *functions*?